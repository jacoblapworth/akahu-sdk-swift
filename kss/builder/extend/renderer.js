const { regexDictionary, types } = require('../parser/constants');
const tinycolor = require('tinycolor2');

module.exports = function(handlebars) {
	/**
	 * Extracts the variable properties from the line
	 * returns:
	 *  token name, token value, and optional [ token example ]
	 */
	handlebars.registerHelper('renderTokens', renderTokens);

	/**
	 * Compares the name to see if the container needs to be rendered as inverted.
	 * Similar to the isEqual
	 *
	 * e.g.
	 * {{#isInverted name}}
	 * 		ADD THIS
	 * {{else}} [optional]
	 * 		WILL EXECUTE THIS
	 * {{/isInverted}}
	 */
	handlebars.registerHelper('isInverted', isInverted);
};

const testString = 'To understand what recursion is, you must first understand recursion.';


function renderTokens(variables, block) {
	var newBlock = '';
	variables.split('\n').forEach(line => {
		var test;
		const matches = line.match(/[^:]+/g).map(match => match.trim());

		if (matches[1]) {
			this.token = {};
			this.token.name = matches[0];
			this.token.value = matches[1];
			this.token.title = matches[3];
			this.token.description = matches[4];
			if (matches[2] != undefined) {
				this.token.example = getExample(matches[2], this.token);
			}
			newBlock += block.fn(this);
		}
	});
	return newBlock;
}

function getExample(type, token) {
	switch(type) {
		case types.color:
			const codeTextClass = tinycolor(token.value).isDark() ?
				'ds-colour-code-light'
				: null;
			return `
					<div class="ds-colour-example" style="background-color:${token.value}">
						<code class="${codeTextClass}">${token.value}</code>
						<code class="${codeTextClass}">${token.name}</code>
					</div>`;
		case types.border:
			return `
				<div class="ds-colour-example" style="border:${token.value}">
					<code>${token.name}</code>
				</div>`;
		case types.shadow:
			return `
					<div class="ds-colour-example" style="box-shadow:${token.value}">
						<code>${token.name}</code>
					</div>`;
		case types.fontSize:
			return  `<span class="xui-text-truncated" style="font-size:${token.value};">Aa</span>`;
		case types.fontWeight:
			return `<span style="font-weight:${token.value};">Aa</span>`;
		case types.lineHeight:
			return `<span style="line-height:${token.value}">${testString}</span>`;
		case types.spacing:
			return `<div class="ds-spacing-example" style="height:${token.value};width:${token.value};"></div>`;
		default:
			return '';
	}
}

function isInverted(name, options) {
	return regexDictionary.inverted.test(name) ? options.fn(this) : options.inverse(this);
}
