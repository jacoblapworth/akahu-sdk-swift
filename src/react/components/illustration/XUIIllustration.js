import cn from 'classnames';
import PropTypes from 'prop-types';
import React from 'react';

import { baseClass, sizeClasses } from './private/constants';

const XUIIllustration = ({
  alt,
  className,
  height,
  padding,
  size,
  qaHook,
  style,
  ...spreadProps
}) => {
  const classes = cn(className, baseClass, size && sizeClasses[size]);

  const composedStyle = {
    ...style,
    height: typeof height === 'number' ? `${height}px` : height,
    padding: typeof padding === 'number' ? `${padding}px` : padding,
  };

  return (
    <img
      {...spreadProps}
      alt={alt}
      className={classes}
      data-automationid={qaHook}
      style={composedStyle}
    />
  );
};

XUIIllustration.propTypes = {
  /**
   * The alternative text for the `img` element.
   *
   * Defaults to an empty string (`alt=""`) which indicates that this image is not a key part of the
   * content (decorative), and that non-visual browsers may omit it from rendering.
   *
   * For more information [visit the MDN page for the Image Embed
   * element](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/img#Attributes).
   */
  alt: PropTypes.string,
  className: PropTypes.string,
  /**
   * We recommend using one of the provided sizes, but you can use the `height` prop if you need to
   * set a custom height.
   *
   * To set the height of the illustration, provide the desired height in pixels as a `number`, or
   * as a `string` with a CSS unit prefix.
   *
   * e.g. `100%`
   */
  height: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
  /**
   * The padding for the illustration.
   *
   * To set the padding for the illustration, provide the desired padding in pixels as a `number`,
   * or as a string with a CSS unit prefix (e.g. `10%`).
   */
  padding: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
  qaHook: PropTypes.string,
  /**
   * The size of the illustration.
   *
   * The available sizes are `small`, `medium`, and `large`.
   */
  size: PropTypes.oneOf(Object.keys(sizeClasses)),
  /**
   * The url for the illustration.
   *
   * You can find the link for your illustration in [the Illustration
   * repository](https://github.dev.xero.com/UXE/illustrations/tree/master/src/).
   */
  src: PropTypes.string.isRequired,
  /**
   * Optional styles to be added to the `img` element. Styles built into the component will take
   * priority.
   */
  style: PropTypes.object,
};

XUIIllustration.defaultProps = {
  alt: '',
};

export default XUIIllustration;
