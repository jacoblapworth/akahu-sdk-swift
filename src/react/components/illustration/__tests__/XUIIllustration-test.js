import React from 'react';
import Enzyme, { mount, render } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import XUIIllustration from '../XUIIllustration';
import renderer from 'react-test-renderer';

Enzyme.configure({ adapter: new Adapter() });

const illustrationURL = 'illustration.svg';

describe('XUIIllustration', () => {
  it('Should render with the correct classes for small, medium, and large sizes', function() {
    const wrapper = renderer.create(
      <div>
        <XUIIllustration src={illustrationURL} size="small" />
        <XUIIllustration src={illustrationURL} size="medium" />
        <XUIIllustration src={illustrationURL} size="large" />
      </div>,
    );

    expect(wrapper).toMatchSnapshot();
  });

  it('Should render with any additional classes provided through the className prop', () => {
    const wrapper = renderer.create(
      <XUIIllustration src={illustrationURL} className={'classyMcClassFace'} />,
    );

    expect(wrapper).toMatchSnapshot();
  });

  describe('Custom height and padding values', () => {
    const theories = [
      { props: { height: 100 }, expectedStyling: { height: '100px' } },
      { props: { padding: 10 }, expectedStyling: { padding: '10px' } },
      {
        props: { height: '100px', padding: '10px' },
        expectedStyling: { height: '100px', padding: '10px' },
      },
      {
        props: { size: 'small', padding: '10px' },
        expectedStyling: { className: 'xui-illustration-small', padding: '10px' },
      },
      {
        props: { size: 'small', height: '100px', padding: '10px' },
        expectedStyling: {
          className: 'xui-illustration-small',
          height: '100px',
          padding: '10px',
        },
      },
      {
        props: { height: 'invalid height', padding: 'invalid padding' },
        expectedStyling: {
          height: '',
          padding: '',
        },
      },
    ];

    theories.map(theory => {
      it(`Size: ${String(theory.props.size).padEnd(10)} | Height: ${String(
        theory.props.height,
      ).padEnd(10)} | Padding: ${theory.props.padding}`, () => {
        const wrapper = mount(<XUIIllustration src={illustrationURL} {...theory.props} />);

        const element = wrapper.getDOMNode();

        if (theory.expectedStyling.className) {
          expect(element.classList.contains(theory.expectedStyling.className)).toBe(true);
        }
        expect(getComputedStyle(element).height).toEqual(theory.expectedStyling.height || '');
        expect(getComputedStyle(element).padding).toEqual(theory.expectedStyling.padding || '');
      });
    });
  });
});
