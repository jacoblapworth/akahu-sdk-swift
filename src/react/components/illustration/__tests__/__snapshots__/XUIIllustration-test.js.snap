// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`XUIIllustration Should render with any additional classes provided through the className prop 1`] = `
<img
  alt=""
  className="classyMcClassFace xui-illustration"
  src="illustration.svg"
  style={
    Object {
      "height": undefined,
      "padding": undefined,
    }
  }
/>
`;

exports[`XUIIllustration Should render with the correct classes for small, medium, and large sizes 1`] = `
<div>
  <img
    alt=""
    className="xui-illustration xui-illustration-small"
    src="illustration.svg"
    style={
      Object {
        "height": undefined,
        "padding": undefined,
      }
    }
  />
  <img
    alt=""
    className="xui-illustration xui-illustration-medium"
    src="illustration.svg"
    style={
      Object {
        "height": undefined,
        "padding": undefined,
      }
    }
  />
  <img
    alt=""
    className="xui-illustration xui-illustration-large"
    src="illustration.svg"
    style={
      Object {
        "height": undefined,
        "padding": undefined,
      }
    }
  />
</div>
`;
